keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+
+QUESTION garbage collection, cluster and namespaced scoped objects. Which will be the best scope for these new objects? If we can make that IPAddresses are owned by the Services that allocate them, the garbage collector when deleting a Service will take care of the release????
+
+<<[/UNRESOLVED]>>
+```
+
+### NodePort
+
+TBD
+
+
+### Test Plan
+
+* Unit tests
+
+TBD
+
+* Integration Tests
+
+TBD
+
+* E2E tests 
+
+
+### Graduation Criteria
+
+The change must be transparent to the users, so there should not be any performance degradation
+or Services breaking changes. 
+
+### Upgrade / Downgrade Strategy
+
+The service-subnet and node-port range will keep using the flags for the initial configuration, it can
+happen that the service subnet or node-port range are modified after the initial configuration, in that
+case they will use the same value as configured by the flags.
+
+### Version Skew Strategy
+
+That apisever control-plane will keep using repair loops to sync current service IP allocated with
+the corresponding allocator, that guarantee that the skeweved apiservers work because they have the
+same source of truth.
+
+```
+<<[UNRESOLVED detect apiservers with different version ]>>
+
+Any update on the IPRange allocated object will not be allowed if there are apiservers without the
+same version
+QUESTION Is this even possible?
+
+<<[/UNRESOLVED]>>
+```
+
+## Implementation History
+
+### Phase I
+
+The first phase consists in introduce the new API objects as a dropin replacement of current bitmap allocators.
+This can be done just [implementing the `Allocator.Interface`](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/allocator.go) using the new API objects:
+
+```go
+// Interface manages the allocation of IP addresses out of a range. Interface
+// should be threadsafe.
+type Interface interface {
+	Allocate(net.IP) error
+	AllocateNext() (net.IP, error)
+	Release(net.IP) error
+	ForEach(func(net.IP))
+	CIDR() net.IPNet
+
+	// For testing
+	Has(ip net.IP) bool
+}
+```
+
+There is actually a working prototype with this approach in https://github.com/kubernetes/kubernetes/pull/98280
+
+### Phase 2
+
+The next phase consists in simplifying the Serice REST API leveraging the new API changes introduced in Phase I.
+
+This buils on top of Tim Hockin changes:
+
+https://github.com/kubernetes/kubernetes/pull/96684
+https://github.com/kubernetes/kubernetes/pull/95666
+
+
+With the new hooks and the new API fields, we can use API machinery tools and primitives to deal
+with the resource allocation and simplify the Services REST Strategy.
+
+
+```
+<<[UNRESOLVED allocated resources deletion ]>>
+
+Allocated resources can be owned by the services, so the Garbage Collector guarantees the release of the
+allocated resources with the 
+
+<<[/UNRESOLVED]>>
+```
+
+## Drawbacks
+
+Compared to a bitmap this design takes more space per reservation,
```

Comment:
The real drawback is that the number of etcd objects required for this scales at O(N) where N is the number of services.  That seems reasonable (the largest number of services per cluster I have seen peaks at about 15-30k).  etcd memory use is proportional to number of keys, but since pod counts tend to dominate service counts I think this is a reasonable tradeoff.  We still budget 1-2 million keys as a reasonable upper bound for number of etcd keys in a cluster, and this doesn't make that substantially worse (<10% impact).  The objects described here are smaller than most other Kube objects (especially pods), so again the next factor of etcd storage size is still reasonable.